### PR TITLE
[FSTORE-855] Query object refactoring and docs

### PR DESCRIPTION
**Pull request by tdoehmen**
This PR adds/fixes/changes...
- make .features property return all features within the query
- added .feature property to feature view
- added __repr__ for training dataset features for better printout
- added lookup functions needed by transformation functions engine, feature view engine, and flyingduck, removing code duplications
- added .featuregroups property return all featuregroups within the query
- added __getitem__, __getattr__ to achieve same behaviour as featuregroup object
- fixed query.append_feature
- fixed tests

JIRA Issue: FSTORE-855

Priority for Review: -

Related PRs: -

**How Has This Been Tested?**

- [x] Unit Tests
- [ ] Integration Tests
- [x] Manual Tests on VM

**Checklist For The Assigned Reviewer:**

```
- [ ] Checked if merge conflicts with master exist
- [ ] Checked if stylechecks for Java and Python pass
- [ ] Checked if all docstrings were added and/or updated appropriately
- [ ] Ran spellcheck on docstring
- [ ] Checked if guides & concepts need to be updated
- [ ] Checked if naming conventions for parameters and variables were followed
- [ ] Checked if private methods are properly declared and used
- [ ] Checked if hard-to-understand areas of code are commented
- [ ] Checked if tests are effective
- [ ] Built and deployed changes on dev VM and tested manually
- [x] (Checked if all type annotations were added and/or updated appropriately)
```

### CODE REVIEW

**Review by bubriks — COMMENTED** [review]
some quick comments, will have to check it in more detail later

File: python/hsfs/constructor/query.py [hunk 1]
```diff
@@ -23,9 +23,28 @@
...
+        "Feature name '{}' is ambiguous and exists in more than one feature group. "
+        "Consider accessing the feature through the feature group object when specifying the query."
+    )
+    ERROR_MESSAGE_FEATURE_NOT_FOUND = (
+        "Feature name '{}' could not found be found in query."
+    )
+    ERROR_MESSAGE_FEATURE_NOT_FOUND_FG = (
+        "Feature name '{}' could not be found in "
+        "any of the featuregroups in this query."
+    )
```
[the remark this answers is not in the document]
Comment (tdoehmen) [reply python/hsfs/constructor/query.py:46]: Yep, makes it easier for testing, also not too bad to have the message there as static variable.

File: python/hsfs/core/transformation_function_engine.py [hunk 2]
```diff
@@ -100,27 +99,6 @@ def attach_transformation_fn(training_dataset_obj=None, feature_view_obj=None):
             target_obj = training_dataset_obj  # todo why provide td and fv just to convert to target_obj?
```
[the remark this answers is not in the document]
Comment (tdoehmen) [reply python/hsfs/core/transformation_function_engine.py:99]: Out of scope, not part of this PR

File: python/hsfs/constructor/query.py [hunk 3]
```diff
@@ -50,6 +69,95 @@ def __init__(
...
+                    feature_entry
+                ]
+            self._query_feature_list.append(feature_entry)
+
+    def _populate_collections(self):
+        self._featuregroups = {self._left_feature_group}
+        self._query_features = {}
+        self._query_feature_list = []
+        self._featuregroup_features = {}
+        self._filters = self._filter
```
[the remark this answers is not in the document]
Comment (tdoehmen) [reply python/hsfs/constructor/query.py:94]: I'm not sure if you wean that, but we (re-)initialize the lists/dicts here so they can be filled from scratch? We maintain a list and a map of features, to cater to different access patterns. (list for `.features` and dict for `get_feature` etc).

File: python/hsfs/constructor/query.py [hunk 4]
```diff
@@ -50,6 +69,95 @@ def __init__(
...
+                self._add_to_collection(
+                    feat, join_obj.prefix, join_obj.query._left_feature_group
+                )
+            for feat in join_obj.query._left_feature_group.features:
+                self._add_to_collection(
+                    feat,
+                    join_obj.prefix,
+                    join_obj.query._left_feature_group,
+                    query_feature=False,
+                )
```
[the remark this answers is not in the document]
Comment (tdoehmen) [reply python/hsfs/constructor/query.py]: Right, I added another funtion for that.

File: python/hsfs/constructor/query.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (bubriks) [reply python/hsfs/constructor/query.py:100]: oh no here I meant specifically this line: `self._filters = self._filter`

[the remark this answers is not in the document]
Comment (bubriks) [reply python/hsfs/constructor/query.py:100]: Also, I get why you want all of these variables. But the problem (at least to me) is that it is hard to follow what goes into each one and why. I believe this will cause future issues, not only consistency (what you expect to be on the map might not be there even though it is in the list) but also even minor changes will require potentially a lot of debugging (for example someone decided to change the order of tuple, etc.).

[the remark this answers is not in the document]
Comment (tdoehmen) [reply python/hsfs/constructor/query.py:100]: Regarding the filters - every join can contain an additional filter and I collect them all across the joins, starting with the filter in the root query object

[the remark this answers is not in the document]
Comment (tdoehmen) [reply python/hsfs/constructor/query.py:100]: I guess the hard to follow part is also partly the fliside of reducing code duplication :) 
I'm open to suggestions on how to make it easier to follow, but I don't share the opinion that it the double bookkeeping or tuple access by index will inevitably lead to issues in the future :) That's also what we have code review for

**Review by bubriks — APPROVED** [review]

File: python/hsfs/constructor/query.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (tdoehmen) [reply python/hsfs/constructor/query.py]: @bubriks ended up making the change. Can you have a second look?

File: python/hsfs/constructor/query.py [hunk 5]
```diff
@@ -520,31 +530,172 @@ def left_feature_group_start_time(self, left_feature_group_start_time):
...
+        featuregroup_features = {}
+        for feat in self._left_feature_group.features:
+            featuregroup_features[feat.name] = featuregroup_features.get(
+                feat.name, []
+            ) + [self._left_feature_group]
+        for join_obj in self.joins:
+            for feat in join_obj.query._left_feature_group.features:
+                featuregroup_features[feat.name] = featuregroup_features.get(
+                    feat.name, []
+                ) + [join_obj.query._left_feature_group]
```
[the remark this answers is not in the document]
Comment (bubriks) [reply python/hsfs/constructor/query.py:577]: something similar could also be done for assigning `query_features` in `_get_feature_by_name`.

[the remark this answers is not in the document]
Comment (tdoehmen) [reply python/hsfs/constructor/query.py:577]: It's a different logic, because the `features` attribute only contains features that were selected in the "select" part of the query. But filters can be defined on any of the columns in the underlying featuregroups, therefore flyingduck needs this list here to figure out to which featuregroup the features in the filter belong (we cannot rely on the feature.featuregroup_id because we allow this type of statement: `fg.select_all(Feature("name") > 10)` where the Feature object in not bound to a featuregroup, hence the `.featuregroup_id` is empty.
